### PR TITLE
telegram-desktop: build on Darwin

### PR DESCRIPTION
**Pull request by ilya-fedin**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin (with the help of GitHub Actions)
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by NickCao** [comment]
Looks good to me, did you test both aarch64-darwin and x86_64-darwin?

**Comment by ilya-fedin** [comment]
No, I don't have an Apple device, I only built via GitHub actions on x86_64-darwin. ofborg seems to have almost built both but got a timeout.

**Comment by NickCao** [comment]
> No, I don't have an Apple device, I only built via GitHub actions on x86_64-darwin. ofborg seems to have almost built both but got a timeout.

Well then let's wait for some test result from real apple users.

**Comment by ilya-fedin** [comment]
@NickCao have you asked someone or do you rely on someone would notice the PR? I guess in the second case it could be never...

**Comment by NickCao** [comment]
> @NickCao have you asked someone or do you rely on someone would notice the PR? I guess in the second case it could be never...

I called for help in https://t.me/nixos_zhcn.

**Comment by ilya-fedin** [comment]
I asked at #macos:nixos.org in Matrix, too.

**Comment by Cu3PO42** [comment]
It builds and launches on aarch64-darwin.

File: pkgs/applications/networking/instant-messengers/telegram/telegram-desktop/macos.patch [hunk 1]
```diff
@@ -0,0 +1,69 @@
+diff --git a/Telegram/SourceFiles/platform/mac/overlay_widget_mac.mm b/Telegram/SourceFiles/platform/mac/overlay_widget_mac.mm
+index ef544803e9..69c61b3139 100644
+--- a/Telegram/SourceFiles/platform/mac/overlay_widget_mac.mm
++++ b/Telegram/SourceFiles/platform/mac/overlay_widget_mac.mm
+@@ -97,9 +97,11 @@ void MacOverlayWidgetHelper::updateStyles(bool fullscreen) {
+ 	[window setTitleVisibility:NSWindowTitleHidden];
+ 	[window setTitlebarAppearsTransparent:YES];
+ 	[window setStyleMask:[window styleMask] | NSWindowStyleMaskFullSizeContentView];
++#if 0
```
[the remark this answers is not in the document]
Comment (ilya-fedin) [reply pkgs/applications/networking/instant-messengers/telegram/telegram-desktop/macos.patch:9]: The patch will be then just bigger :thinking:

[the remark this answers is not in the document]
Comment (SharzyL) [reply pkgs/applications/networking/instant-messengers/telegram/telegram-desktop/macos.patch:9]: A patch of removing lines would be a little bit smaller I think. But not a big problem hence completely ok to keep the patch as what it is now.

[the remark this answers is not in the document]
Comment (ilya-fedin) [reply pkgs/applications/networking/instant-messengers/telegram/telegram-desktop/macos.patch:9]: You can see that it skips some lines e.g. here
```
+#if 0
 	if (@available(macOS 12.0, *)) {
 		[center
 			addObserver: _observer
@@ -145,6 +146,7 @@ BatterySaving::BatterySaving(Fn<void()> changedCallback) {
 			name: NSProcessInfoPowerStateDidChangeNotification
 			object: nil];
 	}
+#endif
```
It won't be able to skip when removing the lines.

[the remark this answers is not in the document]
Comment (SharzyL) [reply pkgs/applications/networking/instant-messengers/telegram/telegram-desktop/macos.patch:9]: Oh I see, that's true.

**Review by NickCao — APPROVED** [review]

**Comment by ilya-fedin** [comment]
Looks like I removed meta.platforms accidentally instead of changing it to platforms.all :see_no_evil:

**Pull request merged**